lessons/python_packages/lesson.md

```diff
@@ -0,0 +1,258 @@
+# Creating Packages in Python
+
+- Presenter: John Ladan
+- Sources: Elise Olson, [The Hitchhiker's Guide to Packaging](http://the-hitchhikers-guide-to-packaging.readthedocs.org/en/latest/quickstart.html), [*Bill Mills*](https://github.com/BillMills/pythonPackageLesson)
+- Research Field: any
+- Topic: A quick introduction to creating packages in Python
+
+## Why are we doing this?
+
+One of the most powerful things about coding for the sciences is that it costs
+nothing to re-use code we've written in the past, allowing us to build on
+past work rather than starting over every project or paper.
+
+However, one practice we see again and again is copying and pasting code from one project into another. Sometimes it will just be a function, other times (*cough*MATLAB*cough*) it's files.
+
+Assembling
+code in packages makes it really easy to re-use old code: all the scripts and functions end up in a central location and can be called and imported from anywhere on the computer -
+just like the famous packages `numpy` or `matplotlib`.
+
+## Setting up
+
+In this tutorial, we'll walk through the creation of a simple package from some code that's just lying around. You can view and clone the [demo package](https://github.com/jladan/package_demo). For an unpackaged state, checkout the `unpackaged` tag.
+
+## 1. The basics
+
+The most basic directory structure for a Python package looks like this:
+
+```
+project
+|
+|__ setup.py
+|
+|__ myPackage
+     |
+     |__ somePython.py
+     |__ __init__.py
+```
+
+But at the moment, we've just got some flat files.
+
+```
+project
+|
+|__ norms.py
+|__ metrics.py
+```
+
+So, the first step is to move files around. First comes the hardest part: choosing a package name. I'll call mine `measure`. Create a directory with that name, and move the python files in there.
+
+```
+project
+|
+|__ measure
+     |__ norms.py
+     |__ metrics.py
+```
+
+There is one more crucial file: `__init__.py` lets the Python interpreter know that there are importable modules in this directory. This is the script that gets run when you execute `import measure`. For more about what you can do with modules, you can see the (Python docs)[https://docs.python.org/3/tutorial/modules.html]. After adding `__init__.py`, the project directory should be
+
+```
+project
+|
+|__ measure
+     |__ __init__.py
+     |__ norms.py
+     |__ metrics.py
+```
+
+
+## 2. setup.py
+
+At this point, the library can be imported if we're in the same directory, but it isn't a package. To let setuptools and `pip` know how to handle it, we need to add the `setup.py` file.
+
+A very basic version of `setup.py` is
+
+```
+from setuptools import setup
+
+setup(
+    # Whatever arguments you need/want
+)
+```
+
+If you were to run that file, you'd get a whole bunch of warnings, and nothing would actually get packaged. As a bare minimum, here is our `setup.py` for it to work.
+
+```
+from setuptools import setup
+
+setup(
+    # Needed to silence warnings
```

Comment:
No, it would create the distribution file titled "UNKOWN-0.0.0". I tested that specifically. All the contents are actually in the dist file.  It also installed. Not naming your package is pretty horrible, though.